base/accumulate.jl

```diff
@@ -0,0 +1,450 @@
+
+length(acc::Accumulate) = length(acc.iter)
+size(acc::Accumulate)   = size(acc.iter)
+axes(acc::Accumulate)   = axes(acc.iter)
+ndims(acc::Accumulate)  = ndims(acc.iter)
+
+start(acc::Accumulate) = (@_inline_meta; (start(acc.iter), acc.v0))
+done(acc::Accumulate, accstate) = (@_inline_meta; done(acc.iter, accstate[1]))
+function next(acc::Accumulate, accstate)
+    @_inline_meta
+    itrstate, accval = accstate
+    val, itrstate = next(acc.iter, itrstate)
+    if accval === uninitialized
+        accval = reduce_first(acc.op, val)
+        return accval, (itrstate, accval)
+    else
+        accval = acc.op(accval, val)
+        return accval, (itrstate, accval)
+    end
+end
+
+accumulate(op, itr) = collect(Accumulate(op, itr))
+accumulate(op, v0, itr) = collect(Accumulate(op, v0, itr))
+
+
+# TODO
+# below was copied directly from old code in multidimensional.jl
+
+# see discussion in #18364 ... we try not to widen type of the resulting array
+# from cumsum or cumprod, but in some cases (+, Bool) we may not have a choice.
+rcum_promote_type(op, ::Type{T}, ::Type{S}) where {T,S<:Number} = promote_op(op, T, S)
+rcum_promote_type(op, ::Type{T}) where {T<:Number} = rcum_promote_type(op, T,T)
+rcum_promote_type(op, ::Type{T}) where {T} = T
+
+# handle sums of Vector{Bool} and similar.   it would be nice to handle
+# any AbstractArray here, but it's not clear how that would be possible
+rcum_promote_type(op, ::Type{Array{T,N}}) where {T,N} = Array{rcum_promote_type(op,T), N}
+
+# accumulate_pairwise slightly slower then accumulate, but more numerically
+# stable in certain situations (e.g. sums).
+# it does double the number of operations compared to accumulate,
+# though for cheap operations like + this does not have much impact (20%)
+function _accumulate_pairwise!(op::Op, c::AbstractVector{T}, v::AbstractVector, s, i1, n)::T where {T,Op}
+    @inbounds if n < 128
+        s_ = v[i1]
+        c[i1] = op(s, s_)
+        for i = i1+1:i1+n-1
+            s_ = op(s_, v[i])
+            c[i] = op(s, s_)
+        end
+    else
+        n2 = n >> 1
+        s_ = _accumulate_pairwise!(op, c, v, s, i1, n2)
+        s_ = op(s_, _accumulate_pairwise!(op, c, v, op(s, s_), i1+n2, n-n2))
+    end
+    return s_
+end
+
+function accumulate_pairwise!(op::Op, result::AbstractVector, v::AbstractVector) where Op
+    li = linearindices(v)
+    li != linearindices(result) && throw(DimensionMismatch("input and output array sizes and indices must match"))
+    n = length(li)
+    n == 0 && return result
+    i1 = first(li)
+    @inbounds result[i1] = v1 = v[i1]
+    n == 1 && return result
+    _accumulate_pairwise!(op, result, v, v1, i1+1, n-1)
+    return result
+end
+
+function accumulate_pairwise(op, v::AbstractVector{T}) where T
+    out = similar(v, rcum_promote_type(op, T))
+    return accumulate_pairwise!(op, out, v)
+end
+
+function cumsum!(out, v::AbstractVector, dim::Integer)
+    # we dispatch on the possibility of numerical stability issues
```

Comment:
on the possibility of roundoff errors

(Again, this is misusing the term "numerical stability".  Even naive summation is backwards stable, it is just less accurate than pairwise summation.)